File: src/core/gfx/d3d_device.cc
#include "core/gfx/d3d_device.h"

#include "core/common/errors.h"

#include <winrt/Windows.Graphics.DirectX.Direct3D11.h>
#include <windows.graphics.directx.direct3d11.interop.h>

#include <d3d11.h>

namespace wincap {

namespace {

Microsoft::WRL::ComPtr<IDXGIAdapter4> PickAdapter(LUID preferred) {
    Microsoft::WRL::ComPtr<IDXGIFactory6> factory;
    WINCAP_THROW_IF_FAILED("d3d_device",
        CreateDXGIFactory2(0, IID_PPV_ARGS(factory.GetAddressOf())));

    if (preferred.LowPart != 0 || preferred.HighPart != 0) {
        Microsoft::WRL::ComPtr<IDXGIAdapter1> a1;
        for (UINT i = 0; factory->EnumAdapters1(i, a1.ReleaseAndGetAddressOf()) != DXGI_ERROR_NOT_FOUND; ++i) {
            DXGI_ADAPTER_DESC1 desc{};
            if (FAILED(a1->GetDesc1(&desc))) continue;
            if (desc.AdapterLuid.LowPart == preferred.LowPart &&
                desc.AdapterLuid.HighPart == preferred.HighPart) {
                Microsoft::WRL::ComPtr<IDXGIAdapter4> a4;
                if (SUCCEEDED(a1.As(&a4))) return a4;
            }
        }
    }

    Microsoft::WRL::ComPtr<IDXGIAdapter1> a1;
    WINCAP_THROW_IF_FAILED("d3d_device",
        factory->EnumAdapterByGpuPreference(0, DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE,
                                            IID_PPV_ARGS(a1.GetAddressOf())));
    Microsoft::WRL::ComPtr<IDXGIAdapter4> a4;
    WINCAP_THROW_IF_FAILED("d3d_device", a1.As(&a4));
    return a4;
}

} // namespace

void D3DDevice::Create(LUID preferred_luid) {
    adapter_ = PickAdapter(preferred_luid);

    UINT flags = D3D11_CREATE_DEVICE_BGRA_SUPPORT |
                 D3D11_CREATE_DEVICE_VIDEO_SUPPORT;
#ifndef NDEBUG
    // Debug layer is opt-in; only enable when the SDK layer is installed.
    // flags |= D3D11_CREATE_DEVICE_DEBUG;
#endif

    const D3D_FEATURE_LEVEL levels[] = {
        D3D_FEATURE_LEVEL_11_1,
        D3D_FEATURE_LEVEL_11_0,
    };

    Microsoft::WRL::ComPtr<ID3D11Device>        base_device;
    Microsoft::WRL::ComPtr<ID3D11DeviceContext> base_context;
    D3D_FEATURE_LEVEL got_level{};

    WINCAP_THROW_IF_FAILED("d3d_device",
        D3D11CreateDevice(adapter_.Get(), D3D_DRIVER_TYPE_UNKNOWN, nullptr,
                          flags, levels, _countof(levels), D3D11_SDK_VERSION,
                          base_device.GetAddressOf(), &got_level,
                          base_context.GetAddressOf()));

    WINCAP_THROW_IF_FAILED("d3d_device", base_device.As(&device_));
    WINCAP_THROW_IF_FAILED("d3d_device", base_context.As(&context_));

    // Multithread-protect the immediate context — WGC FrameArrived runs on
    // an MTA worker and may touch the same context as the consumer thread.
    Microsoft::WRL::ComPtr<ID3D10Multithread> mt;
    if (SUCCEEDED(context_.As(&mt))) {
        mt->SetMultithreadProtected(TRUE);
    }

    // Create the WinRT projection (IDirect3DDevice) used by WGC.
    Microsoft::WRL::ComPtr<IDXGIDevice> dxgi_device;
    WINCAP_THROW_IF_FAILED("d3d_device", device_.As(&dxgi_device));

    using winrt::Windows::Graphics::DirectX::Direct3D11::IDirect3DDevice;
    Microsoft::WRL::ComPtr<::IInspectable> inspectable;
    WINCAP_THROW_IF_FAILED("d3d_device",
        CreateDirect3D11DeviceFromDXGIDevice(dxgi_device.Get(), inspectable.GetAddressOf()));

    // Hand off ownership to a raw void*; consumer reinterprets to
    // winrt::Windows::Graphics::DirectX::Direct3D11::IDirect3DDevice.
    winrt_device_ = inspectable.Detach();
}

void D3DDevice::Reset() noexcept {
    if (winrt_device_) {
        static_cast<::IUnknown*>(winrt_device_)->Release();
        winrt_device_ = nullptr;
    }
    context_.Reset();
    device_.Reset();
    adapter_.Reset();
}

} // namespace wincap

File: crates/wincap-core/src/d3d_device.rs
use windows::core::Interface;
use windows::Win32::Graphics::Direct3D::*;
use windows::Win32::Graphics::Direct3D11::*;
use windows::Win32::Graphics::Dxgi::*;
use windows::Win32::Foundation::{HMODULE, LUID};
use windows::Win32::System::WinRT::Direct3D11::{
    CreateDirect3D11DeviceFromDXGIDevice, IDirect3DDxgiInterfaceAccess,
};
use windows::Graphics::DirectX::Direct3D11::IDirect3DDevice;

use crate::error::{hr_call, WincapError, WincapResult};

pub struct D3DDevice {
    pub device: ID3D11Device5,
    pub context: ID3D11DeviceContext4,
    pub adapter: IDXGIAdapter4,
    winrt_device: IDirect3DDevice,
}

// SAFETY: D3D11 device with multithread protection enabled is thread-safe.
unsafe impl Send for D3DDevice {}
unsafe impl Sync for D3DDevice {}

impl D3DDevice {
    /// Create a D3D11 device. If `preferred_luid` is non-zero, picks the
    /// matching adapter; otherwise uses the high-performance adapter.
    pub fn create(preferred_luid: LUID) -> WincapResult<Self> {
        let adapter = pick_adapter(preferred_luid)?;

        let flags = D3D11_CREATE_DEVICE_BGRA_SUPPORT | D3D11_CREATE_DEVICE_VIDEO_SUPPORT;
        let levels = [D3D_FEATURE_LEVEL_11_1, D3D_FEATURE_LEVEL_11_0];

        let mut base_device = None;
        let mut base_context = None;
        let mut _got_level = D3D_FEATURE_LEVEL_11_0;

        hr_call!("d3d_device", unsafe {
            D3D11CreateDevice(
                &adapter,
                D3D_DRIVER_TYPE_UNKNOWN,
                HMODULE::default(),
                flags,
                Some(&levels),
                D3D11_SDK_VERSION,
                Some(&mut base_device),
                Some(&mut _got_level),
                Some(&mut base_context),
            )
        });

        let base_device = base_device.ok_or_else(|| WincapError::General {
            component: "d3d_device",
            message: "D3D11CreateDevice returned no device".into(),
        })?;
        let base_context = base_context.ok_or_else(|| WincapError::General {
            component: "d3d_device",
            message: "D3D11CreateDevice returned no context".into(),
        })?;

        let device: ID3D11Device5 = hr_call!("d3d_device", base_device.cast());
        let context: ID3D11DeviceContext4 = hr_call!("d3d_device", base_context.cast());

        // Multithread-protect the immediate context.
        if let Ok(mt) = context.cast::<windows::Win32::Graphics::Direct3D10::ID3D10Multithread>() {
            unsafe { mt.SetMultithreadProtected(true) };
        }

        // Create the WinRT projection used by WGC.
        let dxgi_device: IDXGIDevice = hr_call!("d3d_device", device.cast());
        let inspectable =
            hr_call!("d3d_device", unsafe { CreateDirect3D11DeviceFromDXGIDevice(&dxgi_device) });
        let winrt_device: IDirect3DDevice = hr_call!("d3d_device", inspectable.cast());

        Ok(Self {
            device,
            context,
            adapter,
            winrt_device,
        })
    }

    pub fn winrt_device(&self) -> &IDirect3DDevice {
        &self.winrt_device
    }

    /// Get the underlying D3D11 texture from a WinRT IDirect3DSurface.
    pub fn surface_to_texture(
        surface: &windows::Graphics::DirectX::Direct3D11::IDirect3DSurface,
    ) -> WincapResult<ID3D11Texture2D> {
        let access: IDirect3DDxgiInterfaceAccess = hr_call!("d3d_device", surface.cast());
        let texture: ID3D11Texture2D = hr_call!("d3d_device", unsafe { access.GetInterface() });
        Ok(texture)
    }
}

fn pick_adapter(preferred: LUID) -> WincapResult<IDXGIAdapter4> {
    let factory: IDXGIFactory6 =
        hr_call!("d3d_device", unsafe { CreateDXGIFactory2(DXGI_CREATE_FACTORY_FLAGS(0)) });

    // Try to find the preferred LUID adapter.
    if preferred.LowPart != 0 || preferred.HighPart != 0 {
        let mut i = 0u32;
        loop {
            let result: Result<IDXGIAdapter1, _> = unsafe { factory.EnumAdapters1(i) };
            match result {
                Ok(adapter) => {
                    if let Ok(desc) = unsafe { adapter.GetDesc1() } {
                        if desc.AdapterLuid.LowPart == preferred.LowPart
                            && desc.AdapterLuid.HighPart == preferred.HighPart
                        {
                            if let Ok(a4) = adapter.cast::<IDXGIAdapter4>() {
                                return Ok(a4);
                            }
                        }
                    }
                    i += 1;
                }
                Err(_) => break,
            }
        }
    }

    // Fallback: high-performance adapter.
    let adapter: IDXGIAdapter1 = hr_call!("d3d_device", unsafe {
        factory.EnumAdapterByGpuPreference(0, DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE)
    });
    let a4: IDXGIAdapter4 = hr_call!("d3d_device", adapter.cast());
    Ok(a4)
}
